gobblin-runtime/src/main/java/gobblin/runtime/job_catalog/JobCatalogListenersList.java

```diff
@@ -1,154 +1,148 @@
     Preconditions.checkNotNull(addedJob);
-    callbackAllListeners(new AddJobCallback(addedJob));
+    try {
+      _disp.execCallbacks(new JobCatalogCallbackFactory(new AddJobCallback(addedJob)));
+    } catch (InterruptedException e) {
+      getLog().warn("onAddJob interrupted.");
+    }
   }
 
   @Override
   public synchronized void onDeleteJob(JobSpec deletedJob) {
     Preconditions.checkNotNull(deletedJob);
-    callbackAllListeners(new DeleteJobCallback(deletedJob));
+    try {
+      _disp.execCallbacks(new JobCatalogCallbackFactory(new DeleteJobCallback(deletedJob)));
+    } catch (InterruptedException e) {
+      getLog().warn("onDeleteJob interrupted.");
+    }
   }
 
   @Override
   public synchronized void onUpdateJob(JobSpec originalJob, JobSpec updatedJob) {
     Preconditions.checkNotNull(originalJob);
     Preconditions.checkNotNull(updatedJob);
-    callbackAllListeners(new UpdateJobCallback(originalJob, updatedJob));
-  }
-
-  public void callbackAllListeners(Callback callback) {
-    for (JobCatalogListener listener: _listeners) {
-      callbackOneListener(callback, listener);
+    try {
+      _disp.execCallbacks(new JobCatalogCallbackFactory(new UpdateJobCallback(originalJob, updatedJob)));
+    } catch (InterruptedException e) {
+      getLog().warn("onUpdateJob interrupted.");
     }
   }
 
-  public void callbackOneListener(Callback callback, JobCatalogListener listener) {
-    String callbackMsg = null;
-    if (_debugLogEnabled) {
-      callbackMsg = "callback " + callback + " on " + listener;
-      _log.debug("Started: " + callbackMsg);
-    }
+  public void callbackOneListener(Function<JobCatalogListener, Void> callback,
+                                  JobCatalogListener listener) {
     try {
-      callback.invoke(listener);
-    }
-    catch (RuntimeException e) {
-      if (null == callbackMsg) {
-        callbackMsg = "callback " + callback + " on " + listener;
-      }
-      _log.error("FAILED: " + callbackMsg + " ; error : " + e);
-    }
-    if (_debugLogEnabled) {
-      _log.debug("Finished: " + callbackMsg);
+      _disp.execCallbacks(new JobCatalogCallbackFactory(callback), listener);
+    } catch (InterruptedException e) {
+      getLog().warn("callback interrupted: "+ callback);
     }
   }
 
-  @RequiredArgsConstructor
-  public static abstract class Callback {
-    final protected String methodName;
-    final protected JobSpec param1;
-    final protected JobSpec param2;
 
-    @Override
-    public String toString() {
-      StringBuilder sb = new StringBuilder(this.methodName);
-      sb.append('(');
-      sb.append('[');
-      jobSpecToParamString(param1, sb);
-      if (null != param2) {
-        sb.append(']');
-        sb.append(',');
-        jobSpecToParamString(param2, sb);
-      }
-      sb.append(']');
-      sb.append(')');
-      return sb.toString();
+  public static class AddJobCallback extends Callback<JobCatalogListener, Void> {
```

Comment:
Could this be inner static classes of `JobCatalogListener` instead? They are not specific to this class.